stl/src/tzdb.cpp

```diff
@@ -1,15 +1,302 @@
+    if (_Acquire_icu_functions() < _Icu_api_level::_Has_icu_addresses) {
+        _Info->_Err = __std_tzdb_error::_Win_error;
+        return _Info.release();
+    }
+
+    UErrorCode _UErr{};
+    _STD unique_ptr<UEnumeration, decltype(&__icu_uenum_close)> _Enum{
+        __icu_ucal_openTimeZoneIDEnumeration(USystemTimeZoneType::UCAL_ZONE_TYPE_ANY, nullptr, nullptr, &_UErr),
+        &__icu_uenum_close};
+    if (U_FAILURE(_UErr)) {
+        _Info->_Err = __std_tzdb_error::_Icu_error;
+        return _Info.release();
+    }
+
+    // uenum_count may be expensive but is required to pre allocated arrays.
+    int32_t _Num_time_zones = __icu_uenum_count(_Enum.get(), &_UErr);
+    if (U_FAILURE(_UErr)) {
+        _Info->_Err = __std_tzdb_error::_Icu_error;
+        return _Info.release();
+    }
+
+    _Info->_Num_time_zones = static_cast<size_t>(_Num_time_zones);
+    _Info->_Names          = new (_STD nothrow) const char*[_Info->_Num_time_zones];
+    if (_Info->_Names == nullptr) {
+        return nullptr;
+    }
+
+    // init to ensure __std_tzdb_delete_init_info cleanup is valid
+    _STD fill_n(_Info->_Names, static_cast<ptrdiff_t>(_Info->_Num_time_zones), nullptr);
+
+    _Info->_Links = new (_STD nothrow) const char*[_Info->_Num_time_zones];
+    if (_Info->_Links == nullptr) {
+        return nullptr;
+    }
+
+    // init to ensure __std_tzdb_delete_init_info cleanup is valid
+    _STD fill_n(_Info->_Links, static_cast<ptrdiff_t>(_Info->_Num_time_zones), nullptr);
+
+    for (size_t _Name_idx = 0; _Name_idx < _Info->_Num_time_zones; ++_Name_idx) {
+        int32_t _Elem_len{};
+        const auto* _Elem = __icu_uenum_unext(_Enum.get(), &_Elem_len, &_UErr);
+        if (U_FAILURE(_UErr) || _Elem == nullptr) {
+            _Info->_Err = __std_tzdb_error::_Icu_error;
+            return _Info.release();
+        }
+
+        _Info->_Names[_Name_idx] = _Allocate_wide_to_narrow(_Elem, _Elem_len, _Info->_Err);
+        if (_Info->_Names[_Name_idx] == nullptr) {
+            return _Info->_Err != __std_tzdb_error::_Success ? _Info.release() : nullptr;
+        }
+
+        UBool _Is_system{};
+        // FIXME: what is the right size here and is it ok to stack allocate.
+        char16_t _Link_buf[256];
+        const auto _Link_buf_len =
+            __icu_ucal_getCanonicalTimeZoneID(_Elem, _Elem_len, _Link_buf, sizeof(_Link_buf), &_Is_system, &_UErr);
+        if (U_FAILURE(_UErr) || _Link_buf_len == 0) {
+            _Info->_Err = __std_tzdb_error::_Icu_error;
+            return _Info.release();
+        }
+
+        if (_STD u16string_view{_Elem, static_cast<size_t>(_Elem_len)}
+            != _STD u16string_view{_Link_buf, static_cast<size_t>(_Link_buf_len)}) {
+            _Info->_Links[_Name_idx] = _Allocate_wide_to_narrow(_Link_buf, _Link_buf_len, _Info->_Err);
+            if (_Info->_Links[_Name_idx] == nullptr) {
+                return _Info->_Err != __std_tzdb_error::_Success ? _Info.release() : nullptr;
+            }
+        }

```

Comment:
Given our discussion from Tuesday evening, I've been thinking more about what exactly should be in this `links` list.  I know it has to exist because it's a public member of `tzdb`, but we already know it's not going to match exactly what would otherwise be considered "alternate" names by IANA.  Also, since everything that's currently put into the `time_zone_link`s list is also in the `time_zone` list, we'll never search through `links` through a call to `locate_zone()` which is the only way a library function uses `links` I believe.  

We've decided that since we are providing a superset of the `time_zone`s in `zones` that would be returned by the IANA database, it's okay that the contents are different.  Should we do something similar with `links`?  So that if a user asks for a `time_zone_link` that exists in IANA, we can still give them an answer (with the current scheme, I believe we won't be able to give an answer for everything the user might ask for since the "any" names in ICU doesn't correspond to the "alternate" names in IANA).  Thoughts here?

Reply:
When you get to the tests I have included a single example of every type of inaccuracy. All of them are common except for the case when a `time_zone` doesn't exist in any form in the ICU (only 1 recent IANA addition). Sometimes the actual target of the link in the ICU is wrong, often because those links can and do get updated in the IANA to point to different things.

Current implementation has neither `zones` or `links` as a super set. The only way to get a complete super set would be to include no `time_zone_link`s what so ever and have everything be a zone. Some `links` according to the ICU are `time_zones` according to the IANA and visa versa.

I think the current solution is the best choice of a set of bad options. I can't see another option without removing all `links` or keeping a manual map of some kind.

Given that users can only use `locate_zone()` to query and that will search both vectors I think the distinction will only matter when people want 100% accurate name parity with the IANA. With ICU data the user will still get reasonable and legitimate names that maintained and updated over time. And these names will match any other windows tool built on the ICU like the .dotnet implementation.

Reply:
Ah okay, I think I missed the fact that what is put into the `Links` structure is `_Link_buf` (the canonical name), not `_Elem` (from the enumerated time zones) if the canonical name doesn't match.  Just to be clear, when we open the enumeration via `openIDTimeZoneEnumeration` with `UCAL_ZONE_TYPE_ANY`, does the enumeration include canonical names as well?  I'm unclear on whether `UCAL_ZONE_TYPE_ANY` includes all time zones (canonical or otherwise), or whether it's a disjoint categorization from those that might be returned by `getCanonicalTimeZoneID`.

Reply:
`UCAL_ZONE_TYPE_ANY` includes both ICU "links" and "canonical" time_zones in one go. There is not way to get only the "links" hence the use of `getCanonicalTimeZoneID`.